packages/astro/src/vite-plugin-astro/hmr.ts

```diff
@@ -1,62 +1,44 @@
-import path from 'node:path';
-import { appendForwardSlash } from '@astrojs/internal-helpers/path';
 import type { HmrContext } from 'vite';
 import type { Logger } from '../core/logger/core.js';
-import type { CompileAstroResult } from './compile.js';
-import type { CompileMetadata } from './types.js';
 import { frontmatterRE } from './utils.js';
+import type { CompileMetadata } from './types.js';
 
 export interface HandleHotUpdateOptions {
 	logger: Logger;
-	compile: (code: string, filename: string) => Promise<CompileAstroResult>;
-	astroFileToCssAstroDeps: Map<string, Set<string>>;
 	astroFileToCompileMetadata: Map<string, CompileMetadata>;
 }
 
 export async function handleHotUpdate(
 	ctx: HmrContext,
-	{ logger, compile, astroFileToCssAstroDeps, astroFileToCompileMetadata }: HandleHotUpdateOptions
+	{ logger, astroFileToCompileMetadata }: HandleHotUpdateOptions
 ) {
-	const oldCode = astroFileToCompileMetadata.get(ctx.file)?.originalCode;
+	const compileMetadata = astroFileToCompileMetadata.get(ctx.file);
+
+	// If `ctx.file` is part of a CSS dependency of any Astro file, invalidate its `astroFileToCompileMetadata`
+	// so the next transform will re-generate it.
+	for (const [astroFile, compileData] of astroFileToCompileMetadata) {
+		const isUpdatedFileCssDep = compileData.css.some((css) => css.dependencies?.includes(ctx.file));
+		if (isUpdatedFileCssDep) {
+			astroFileToCompileMetadata.delete(astroFile);
+		}
+	}
+
+	// Bail if not Astro file, we only handle them below
+	if (compileMetadata == null || !ctx.file.endsWith('.astro')) return;
```

Comment:
Why do we bail here not before the loop? The loop assumes `ctx.file` is an astro file, but here we bail. 

Also, line 15 can be moved after the loop because we don't use it before the loop.

Reply:
The for-loop doesn't assume `ctx.file` to be an Astro file. It could be any file, hence we check against the list there. Line 15 is called before the loop, because line 22 could have deleted it. We still want to narrow down Astro file HMR even after the for-loop matches it. The for-loop is kinda doing its own handling regarding CSS dependencies HMR.

Reply:
Re line 15, thinking about it more I think you're right too. The only case where line 22 could mis-delete it is when `ctx.file` has a dependency of `ctx.file` itself, which is highly unlikely. I'll move that down then.